File: bin/nga.c
/* Nga ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   Copyright (c) 2008 - 2016, Charles Childers
   Copyright (c) 2009 - 2010, Luke Parrish
   Copyright (c) 2010,        Marc Simpson
   Copyright (c) 2010,        Jay Skeer
   Copyright (c) 2011,        Kenneth Keating
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#define CELL         int32_t
#define IMAGE_SIZE   256*1024
#define ADDRESSES    128
#define STACK_DEPTH  32
#define CELLSIZE     32
enum vm_opcode {
  VM_NOP,  VM_LIT,    VM_DUP,   VM_DROP,    VM_SWAP,   VM_PUSH,
  VM_POP,  VM_JUMP,   VM_CALL,  VM_IF,      VM_RETURN,
  VM_EQ,   VM_NEQ,    VM_LT,    VM_GT,      VM_FETCH,  VM_STORE,
  VM_ADD,  VM_SUB,    VM_MUL,   VM_DIVMOD,  VM_AND,    VM_OR,
  VM_XOR,  VM_SHIFT,  VM_ZRET,  VM_END
};
#define NUM_OPS VM_END + 1
CELL sp, rp, ip;
CELL data[STACK_DEPTH];
CELL address[ADDRESSES];
CELL image[IMAGE_SIZE];
int stats[NUM_OPS];
int max_sp, max_rp;
#define DROP { data[sp] = 0; if (--sp < 0) ip = IMAGE_SIZE; }
#define TOS  data[sp]
#define NOS  data[sp-1]
#define TORS address[rp]
CELL ngaLoadImage(char *imageFile) {
  FILE *fp;
  CELL imageSize;
  long fileLen;

  if ((fp = fopen(imageFile, "rb")) != NULL) {
    fseek(fp, 0, SEEK_END);
    fileLen = ftell(fp) / sizeof(CELL);
    rewind(fp);
    imageSize = fread(&image, sizeof(CELL), fileLen, fp);
    fclose(fp);
  }
  else {
    printf("Unable to find the ngaImage!\n");
    exit(1);
  }
  return imageSize;
}
void ngaPrepare() {
  ip = sp = rp = max_sp = max_rp = 0;

  for (ip = 0; ip < IMAGE_SIZE; ip++)
    image[ip] = VM_NOP;

  for (ip = 0; ip < STACK_DEPTH; ip++)
    data[ip] = 0;

  for (ip = 0; ip < ADDRESSES; ip++)
    address[ip] = 0;

  for (ip = 0; ip < NUM_OPS; ip++)
    stats[ip] = 0;
}
void check_max() {
  if (max_sp < sp)
    max_sp = sp;
  if (max_rp < rp)
    max_rp = rp;
}

void inst_nop() {
}
void inst_lit() {
  sp++;
  ip++;
  TOS = image[ip];
  check_max();
}
void inst_dup() {
  sp++;
  data[sp] = NOS;
  check_max();
}

void inst_drop() {
  DROP
}

void inst_swap() {
  int a;
  a = TOS;
  TOS = NOS;
  NOS = a;
}

void inst_push() {
  rp++;
  TORS = TOS;
  DROP
  check_max();
}

void inst_pop() {
  sp++;
  TOS = TORS;
  rp--;
}

void inst_jump() {
  ip = TOS - 1;
  DROP
}

void inst_call() {
  rp++;
  TORS = ip;
  ip = TOS - 1;
  DROP
  check_max();
}

void inst_if() {
  int a, b, c;
  rp++;
  TORS = ip;
  a = TOS; DROP;  /* False */
  b = TOS; DROP;  /* True  */
  c = TOS; DROP;  /* Flag  */
  if (c != 0)
    ip = b - 1;
  else
    ip = a - 1;
  check_max();
}

void inst_return() {
  ip = TORS;
  rp--;
}

void inst_eq() {
  int a, b;
  a = TOS; DROP;
  b = TOS; DROP;
  if (b == a)
    TOS = -1;
  else
    TOS = 0;
}

void inst_neq() {
  int a, b;
  a = TOS; DROP;
  b = TOS; DROP;
  if (b != a)
    TOS = -1;
  else
    TOS = 0;
}

void inst_lt() {
  int a, b;
  a = TOS; DROP;
  b = TOS; DROP;
  if (b < a)
    TOS = -1;
  else
    TOS = 0;
}

void inst_gt() {
  int a, b;
  a = TOS; DROP;
  b = TOS; DROP;
  if (b > a)
    TOS = -1;
  else
    TOS = 0;
}

void inst_fetch() {
  TOS = image[TOS];
}

void inst_store() {
  image[TOS] = NOS;
  DROP
  DROP
}

void inst_add() {
  NOS += TOS;
  DROP
}

void inst_sub() {
  NOS -= TOS;
  DROP
}

void inst_mul() {
  NOS *= TOS;
  DROP
}

void inst_divmod() {
  int a, b;
  a = TOS;
  b = NOS;
  TOS = b / a;
  NOS = b % a;
}

void inst_and() {
}

void inst_or() {
}

void inst_xor() {
}

void inst_shift() {
}

void inst_zret() {
}

void inst_end() {
}

void ngaProcessOpcode() {
  CELL a, b, c, opcode;
  opcode = image[ip];

  stats[opcode]++;

  printf("%d: %d\n", ip, opcode);

  switch(opcode) {
    case VM_NOP: inst_nop();   break;
    case VM_LIT: inst_lit();   break;
    case VM_DUP: inst_dup();   break;
    case VM_DROP: inst_drop();         break;
    case VM_SWAP: inst_swap();         break;
    case VM_PUSH: inst_push();         break;
    case VM_POP: inst_pop();         break;
    case VM_JUMP: inst_jump();         break;
    case VM_CALL: inst_call();         break;
    case VM_IF: inst_if();         break;
    case VM_RETURN: inst_return(); break;
    case VM_GT:   inst_gt();   break;
    case VM_LT:   inst_lt();   break;
    case VM_NEQ:  inst_neq();  break;
    case VM_EQ:   inst_eq();   break;
    case VM_FETCH: inst_fetch();     break;
    case VM_STORE: inst_store(); break;
    case VM_ADD: inst_add();         break;
    case VM_SUB: inst_sub(); break;
    case VM_MUL: inst_mul(); break;
    case VM_DIVMOD: inst_divmod(); break;
    case VM_AND:
         a = TOS;
         b = NOS;
         DROP
         TOS = a & b;
         break;
    case VM_OR:
         a = TOS;
         b = NOS;
         DROP
         TOS = a | b;
         break;
    case VM_XOR:
         a = TOS;
         b = NOS;
         DROP
         TOS = a ^ b;
         break;
    case VM_SHIFT:
         /* Left -- TODO */
         a = TOS;
         b = NOS;
         DROP
         TOS = b << a;

         /* Right -- TODO */
         a = TOS;
         DROP
         TOS >>= a;
         break;
    case VM_ZRET:
         if (TOS == 0) {
           DROP
           ip = TORS;
           rp--;
         }
         break;
    case VM_END:
         ip = IMAGE_SIZE;
         break;
    default:
         printf("Error: %d opcode encountered\n", opcode);
         exit(1);
         break;
  }
}
void ngaDisplayStats()
{
  int s, i;

  printf("Runtime Statistics\n");
  printf("NOP:     %d\n", stats[VM_NOP]);
  printf("LIT:     %d\n", stats[VM_LIT]);
  printf("DUP:     %d\n", stats[VM_DUP]);
  printf("DROP:    %d\n", stats[VM_DROP]);
  printf("SWAP:    %d\n", stats[VM_SWAP]);
  printf("PUSH:    %d\n", stats[VM_PUSH]);
  printf("POP:     %d\n", stats[VM_POP]);
  printf("JUMP:    %d\n", stats[VM_JUMP]);
  printf("CALL:    %d\n", stats[VM_CALL]);
  printf(":IF      %d\n", stats[VM_IF]);
  printf("RETURN:  %d\n", stats[VM_RETURN]);
  printf(":EQ      %d\n", stats[VM_EQ]);
  printf(":NEQ     %d\n", stats[VM_NEQ]);
  printf(":LT      %d\n", stats[VM_LT]);
  printf(":GT      %d\n", stats[VM_GT]);
  printf("FETCH:   %d\n", stats[VM_FETCH]);
  printf("STORE:   %d\n", stats[VM_STORE]);
  printf("ADD:     %d\n", stats[VM_ADD]);
  printf("SUB:     %d\n", stats[VM_SUB]);
  printf("MUL:     %d\n", stats[VM_MUL]);
  printf("DIVMOD:  %d\n", stats[VM_DIVMOD]);
  printf("AND:     %d\n", stats[VM_AND]);
  printf("OR:      %d\n", stats[VM_OR]);
  printf("XOR:     %d\n", stats[VM_XOR]);
  printf("SHIFT:   %d\n", stats[VM_SHIFT]);
  printf("ZRET:    %d\n", stats[VM_ZRET]);
  printf("END:     %d\n", stats[VM_END]);
  printf("Max sp:  %d\n", max_sp);
  printf("Max rp:  %d\n", max_rp);

  for (s = i = 0; s < NUM_OPS; s++)
    i += stats[s];
  printf("Total opcodes processed: %d\n", i);
}
int main(int argc, char **argv) {
  int wantsStats, i;
  wantsStats = 1;

  ngaPrepare();

  ngaLoadImage("ngaImage");

  for (ip = 0; ip < IMAGE_SIZE; ip++)
    ngaProcessOpcode();

  if (wantsStats == 1)
    ngaDisplayStats();

  for (i = 1; i <= sp; i++)
      printf("%d, ", data[i]);

  printf("\n");

  exit(0);
}

File: incoming/pascal/nga.pas
// ********************************************************
//  Copyright (c) 2016 Rob Judd <judd@ob-wan.com>
//  Based on C version by Charles Childers et al
//  ISC License - see included file LICENSE
// ********************************************************

unit nga;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils;

type Cell = Longint;

type vm_opcode = (
   VM_NOP, VM_LIT, VM_DUP, VM_DROP, VM_SWAP, VM_PUSH, VM_POP,
   VM_JUMP, VM_CALL, VM_CCALL, VM_RET, VM_EQ, VM_NEQ, VM_LT,
   VM_GT, VM_FETCH, VM_STORE, VM_ADD, VM_SUB, VM_MUL, VM_DIVMOD,
   VM_AND, VM_OR, VM_XOR, VM_SHIFT, VM_ZRET, VM_END
);

var
  NUM_OPS : Byte = ord(VM_END) + 1;    // number of opcodes
  IMAGE_SIZE : Cell = 524288;          // memory image (512k x 32 bits)
  ip, ap, sp : Cell;                   // instruction, address & stack pointers
  data : array [1..32] of Cell;        // stack depth
  address : array [1..128] of Cell;    // addresses
  memory : array [1..524288] of Cell;  // image size
  TOS, NOS, TOA : Cell;                // top stack, next stack & top address

implementation

function ngaLoadImage(imageFile : string) : Cell;
var
  f : File;
  sr : TSearchRec;
  fileLen: Cell;
  imageSize : Cell = 0;
begin
  // Does the file exist?
  if FindFirst(imageFile, faAnyFile-faDirectory, sr) = 0 then
  begin
    // Determine number of Cells
    fileLen := sr.Size div sizeof(Cell);
    // Read the file into memory
    Assignfile(f, imageFile);
    try
      BlockRead(f, memory, fileLen, imageSize);
    finally
      CloseFile(f);
    end;
  end
  else
  begin
    writeln('Unable to find ', imageFile + '!' + #13);
    exit();
  end;
  FindClose(sr);
  result := imageSize;
end;

procedure ngaPrepare();
var
  i : Cell;
begin
  // Initialize pointers
  ip := 0;
  ap := 0;
  sp := 0;
  // Initialize arrays
  for i := 1 to length(data) do
    data[i] := ord(VM_NOP);            // normally zero
  for i := 1 to length(address) do
    address[i] := 0;
  for i := 1 to length(memory) do
    memory[i] := 0;
end;

procedure inst_nop();
begin
  // Do nothing
end;

procedure inst_lit();
begin
  inc(ip);
  inc(sp);
  TOS := memory[ip];
end;

procedure inst_dup();
begin
  inc(sp);
  data[sp] := NOS;
end;

procedure inst_drop();
begin
  data[sp] := 0;
  dec(sp);
  if sp < 0 then
    ip := IMAGE_SIZE;
end;

procedure inst_swap();
var a : Cell;
begin
  a := TOS;
  TOS := NOS;
  NOS := a;
end;

procedure inst_push();
begin
  inc(ap);
  TOA := TOS;
  inst_drop();
end;

procedure inst_pop();
begin
  inc(sp);
  TOS := TOA;
  dec(ap);
end;

procedure inst_jump();
begin
  ip := TOS - 1;
  inst_drop();
end;

procedure inst_call();
begin
  inc(ap);
  TOA := ip;
  ip := TOS - 1;
  inst_drop();
end;

procedure inst_ccall();
var
  a, b : Cell;
begin
  a := TOS;
  inst_drop();                         // false
  b := TOS;
  inst_drop();                         // flag
  if b <> 0 then
  begin
    inc(ap);
    TOA := ip;
    ip := a - 1;
  end;
end;

procedure inst_ret();
begin
  ip := TOA;
  dec(ap);
end;

procedure inst_eq();
begin
  if NOS = TOS then
    NOS := -1
  else
    NOS := 0;
  inst_drop();
end;

procedure inst_neq();
begin
  if NOS <> TOS then
    NOS := -1
  else
    NOS := 0;
  inst_drop();
end;

procedure inst_lt();
begin
  if NOS < TOS then
    NOS := -1
  else
    NOS := 0;
  inst_drop();
end;

procedure inst_gt();
begin
  if NOS > TOS then
    NOS := -1
  else
    NOS := 0;
  inst_drop();
end;

procedure inst_fetch();
begin
  case TOS of
    -1 : TOS := sp - 1;
    -2 : TOS := ap;
  else
    TOS := memory[TOS];
  end;
end;

procedure inst_store();
begin
  memory[TOS] := NOS;
  inst_drop();
  inst_drop();
end;

procedure inst_add();
begin
  NOS := NOS + TOS;
  inst_drop();
end;

procedure inst_sub();
begin
  NOS := NOS - TOS;
  inst_drop();
end;

procedure inst_mul();
begin
  NOS := NOS * TOS;
  inst_drop();
end;

procedure inst_divmod();
var
  a, b : Cell;
begin
  a := TOS;
  b := NOS;
  TOS := b div a;
  NOS := b mod a;
end;

procedure inst_and();
begin
  NOS := NOS and TOS;
  inst_drop();
end;

procedure inst_or();
begin
  NOS := NOS or TOS;
  inst_drop();
end;

procedure inst_xor();
begin
  NOS := NOS xor TOS;
  inst_drop();
end;

procedure inst_shift();
var
  x, y : Cell;
  z : Cell = 0;
begin
  x := NOS;
  y := TOS;
  if TOS < 0 then
    NOS := NOS shl (TOS * -1)
  else
  begin
    if (x < 0) and (y > 0) then
      NOS := x shr y or not(not z shr y)
    else
      NOS := x shr y;
  end;
end;

procedure inst_zret();
begin
  if TOS = 0 then
  begin
    inst_drop();
    ip := TOA;
    dec(ap);
  end;
end;

procedure inst_end();
begin
  ip := IMAGE_SIZE;
end;

procedure ngaProcessOpcode(opcode : Cell);
begin
  case opcode of
    1  : inst_nop();
    2  : inst_lit();
    3  : inst_dup();
    4  : inst_drop();
    5  : inst_swap();
    6  : inst_push();
    7  : inst_pop();
    8  : inst_jump();
    9  : inst_call();
    10 : inst_ccall();
    11 : inst_ret();
    12 : inst_eq();
    13 : inst_neq();
    14 : inst_lt();
    15 : inst_gt();
    16 : inst_fetch();
    17 : inst_store();
    18 : inst_add();
    19 : inst_sub();
    20 : inst_mul();
    21 : inst_divmod();
    22 : inst_and();
    23 : inst_or();
    24 : inst_xor();
    25 : inst_shift();
    26 : inst_zret();
    27 : inst_end();
  end;
end;

function ngaValidatePackedOpcodes(opcode : Cell) : Integer;
var
  raw, current : Cell;
  valid : Integer = - 1;               // value for "true" in Unix-land
  i : Integer;
begin
  raw := opcode;
  for i := 1 to 4 do
  begin
    current := raw and $FF;
    if ((current >= 0) and (current < NUM_OPS)) = false then
      valid := 0;
    raw := raw shr 8;
  end;
  result := valid;
end;

procedure ngaProcessPackedOpcodes(opcode : Cell);
var
  raw : Cell;
  i : Byte;
begin
  raw := opcode;
  for i := 1 to 4 do
  begin
    ngaProcessOpcode(raw and $FF);
    raw := raw shr 8;
  end;
end;

// ********************************************************
//  Main program
// ********************************************************
var
  i, opcode, size : Cell;
begin
  ngaPrepare();
  if ParamCount > 0 then
    size := ngaLoadImage(ParamStr(1))
  else
    size := ngaLoadImage('ngaImage');

  for ip := 1 to size do
  begin
    opcode := memory[ip];
    if (ngaValidatePackedOpcodes(opcode)) <> 0 then
      ngaProcessPackedOpcodes(opcode)
    else if ((opcode >= 0) and (opcode < NUM_OPS)) = true then
      ngaProcessOpcode(opcode)
    else
    begin
      writeln('Invalid instruction loaded!' + #13);
      writeln('at offset %d, opcode %d.', ip, opcode);
      exit();
    end;
  end;
  // Screen dump addresses for testing
  for i := 1 to ap do
    writeln(' %d', data[i]);
  writeln(#13);
end.

